pkg/machine/compression/decompress.go

```diff
@@ -1,279 +1,113 @@
 		return err
 	}
-	p, bar := utils.ProgressBar(prefix, stat.Size(), prefix+": done")
-	proxyReader := bar.ProxyReader(f)
-	defer func() {
-		if err := proxyReader.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-	uncompressStream, _, err := compression.AutoDecompress(proxyReader)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := uncompressStream.Close(); err != nil {
-			logrus.Error(err)
-		}
-		if err := output.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
+	defer d.close()
 
-	_, err = io.Copy(output, uncompressStream)
-	p.Wait()
-	return err
-}
+	initMsg := progressBarPrefix + ": " + filepath.Base(decompressedFilePath)
+	finalMsg := initMsg + ": done"
 
-func decompressZip(prefix string, src string, output io.WriteCloser) error {
-	zipReader, err := zip.OpenReader(src)
-	if err != nil {
-		return err
-	}
-	if len(zipReader.File) != 1 {
-		return errors.New("machine image files should consist of a single compressed file")
-	}
-	f, err := zipReader.File[0].Open()
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := f.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-	defer func() {
-		if err := output.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-	size := int64(zipReader.File[0].CompressedSize64)
-	p, bar := utils.ProgressBar(prefix, size, prefix+": done")
-	proxyReader := bar.ProxyReader(f)
-	defer func() {
-		if err := proxyReader.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-	_, err = io.Copy(output, proxyReader)
-	p.Wait()
-	return err
-}
+	p, bar := utils.ProgressBar(initMsg, d.compressedFileSize(), finalMsg)
+	// Wait for bars to complete and then shut down the bars container
+	defer p.Wait()
 
-func decompressGzWithSparse(prefix string, compressedPath *define.VMFile, uncompressedPath string) error {
-	stat, err := os.Stat(compressedPath.GetPath())
-	if err != nil {
-		return err
-	}
+	compressedFileReaderProxy := bar.ProxyReader(compressedFileReader)
+	// Interrupts the bar goroutine. It's important that
+	// bar.Abort(false) is called before p.Wait(), otherwise
+	// can hang.
+	defer bar.Abort(false)
 
-	dstFile, err := os.OpenFile(uncompressedPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, stat.Mode())
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := dstFile.Close(); err != nil {
-			logrus.Errorf("unable to close uncompressed file %s: %q", uncompressedPath, err)
-		}
-	}()
+	var decompressedFileWriter *os.File
 
-	f, err := os.Open(compressedPath.GetPath())
-	if err != nil {
+	if decompressedFileWriter, err = os.OpenFile(decompressedFilePath, decompressedFileFlag, d.compressedFileMode()); err != nil {
+		logrus.Errorf("Unable to open destination file %s for writing: %q", decompressedFilePath, err)
 		return err
 	}
 	defer func() {
-		if err := f.Close(); err != nil {
-			logrus.Errorf("unable to close on compressed file %s: %q", compressedPath.GetPath(), err)
+		if err := decompressedFileWriter.Close(); err != nil && !errors.Is(err, os.ErrClosed) {
```

Comment:
`ErrClosed` indicates a programming bug (analogous to C use-after-free); if it is showing up in practice, it should be traced down and avoided.

Reply:
This code exist in main ([introduced last week](https://github.com/containers/podman/commit/487219d809732e69e052e50ee403a4c35c1a9dc5#diff-18f971938aa35cdc3677d52dffdbc71cc7a67e2ff700a8907508301ccab33216)).

Reply:
Yeah, still a bug, and closing unowned file descriptors in a concurrency-heavy language like Go would be maddening to debug. I can, barely, live with deferring #21772, that’s “only” data corruption, but unpredictable in-process memory corruption is even more dangerous than that.

(As it happens, the Go internal implementation _seems_ to atomically enforce only one `Close`, but the interface spec explicitly says the behavior is undefined, so we can’t rely on that.)

AFAICS this refactor has _already_ fixed the double-closes (and that can be enforced by the type system but not providing a `*Closer` type to the individual decompressor implementations [see elsewhere about the sparse writer `Close` handling]), so just removing this check should be sufficient to resolve.

Reply:
I’m sorry, actually the `io.Closer.Close` specification says effect of repeated calls is undefined, but specifically `os.File.Close` documents that “Close will return an error if it has already been called.” (but, curiously, doesn’t commit to returning `ErrClosed` in this case).

So this is, after all, safe enough.

Reply:
https://github.com/containers/podman/pull/21977 will fix this.